src/main/java/com/imgix/URLHelper.java

```diff
@@ -192,11 +201,29 @@ public static String decodeURIComponent(String s) {
     String result = null;
 
     try {
-      result = URLDecoder.decode(s, "UTF-8");
+      result = URLDecoder.decode(s, UTF_8);
     } catch (UnsupportedEncodingException e) {
       result = s;
     }
 
     return result;
   }
+
+  public static String sanatizePath(String path) {
+    // Strip leading slash first (we'll re-add after encoding)
+    path = path.replaceAll("^/", "");
+
+    if (path.startsWith("http://") || path.startsWith("https://")) {
+      // Use encodeURIComponent to ensure *all* characters are handled,
+      // since it's being used as a path
+      return "/" + URLHelper.encodeURIComponent(path);
+    } else if (path.startsWith("http%3A") || path.startsWith("https%3A")) {
+      // To nothing to the URL being used as path since already encoded
+      return "/" + path;
+    } else {
+      // Use encodeURI if we think the path is just a path,
+      // so it leaves legal characters like '/' and '@' alone
+      return "/" + encodeURI(path);
+    }
+  }
```

Comment:
Did my best to keep in line with logic from [js-core](https://github.com/imgix/js-core/blob/0ad839884b2f133a6dd27b5823c74dcbf0e5b47a/src/index.js#L92-L107). I struggled to avoid having to call this method when initializing the class. Something about the order these methods get called when breaks the signing `MD5` method if I call this inside the `buildURL` method instead.

Reply:
It could be that one function expects there to be or not to be a `/` prefixed to the path at one time or another.

One precondition for MD5 signing that we don't talk about (often enough) is that signing seems to require that the signature base be in the form:

```
TOKEN_STRING + /_PATH + ?_QUERY_STRING
```

You can check assumptions with:

```bash
$ echo -n "FOO123bar/users/1.png?h=300&w=400" | md5
```

Reply:
Good call. Easiest way to fix this is to call `sanitizePath` inside all the URL building methods. But that doesn't seem ideal. What if path gets referenced outside one of those methods?

🤔

Reply:
Just moving this to the top of `buildUrl` did the trick. Not a fan of the solution, given the sequence of steps _has_ to happen in this order for things to work. Issue is the way the signing is implemented.